### ISSUE TITLE
How to achieve 'exactly once' semantics by using an SQL database

### ISSUE THREAD

**Issue by SynergyEvolved** [tulios/kafkajs#1205]
In the docs under manual commits (https://kafka.js.org/docs/1.12.0/consuming#manual-commits) it says:

`Note that you don't have to store consumed offsets in Kafka, but instead store it in a storage mechanism of your own choosing. That's an especially useful approach when the results of consuming a message are written to a datastore that allows atomically writing the consumed offset with it, like for example a SQL database. When possible it can make the consumption fully atomic and give "exactly once" semantics that are stronger than the default "at-least once" semantics you get with Kafka's offset commit functionality.`

Could you please point me to a working example of how this would be achieved?

Many thanks

-Paul

**Comment by toledompm** [comment]
Don't have a public working example, but something like this should work:

``` ts
// DB Repository
const repo = {
  save: (_data) => undefined,
  find: (_data) => undefined,
}


// Consumer
const consumer = kafka.consumer({ groupId: 'my-group' })

await consumer.connect()
await consumer.subscribe({ topic: 'topic-A' })

await consumer.run({
  eachMessage: async ({ topic, partition, message }) => {
    // You need to get an identifier for the message
    // It can be published together with the data like a fakeID
    // Or be generated from groupId, topic, partition and offset
    const id = `${topic}-${partition}-${message.offset}`

    // Query the DB to check if the message is already processed
    const isProcessed = await repo.find(id)
    if (isProcessed) {
      return
    }

    // Process the message

    // Save the message identifier in the DB
    await repo.save(id)
  },
})
```

Gonna try to create an example repo for this, hope it helps

**Comment by OneCricketeer** [comment]
Related #1221 #1252

**Comment by Nevon** [comment]
Exactly-once is a very misunderstood topic. When Confluent says that Kafka supports exactly-once semantics, they mean that in the sense that the _observable outcome_ of a topic being processed is the same whether a message has been consumed once or several times, where the observable outcome is a different topic. This is strictly within the context of a stream consuming from one topic and producing to another.

It does not mean that the message is only ever seen once. It just means that if someone is consuming the output topic, the result will be the same whether the stream processor processed the input message once or several times. This can be achieved using a transactional producer. https://dzone.com/articles/interpreting-kafkas-exactly-once-semantics

What @toledompm describes is not exactly-once, for several reasons. First, what if the group is rebalancing and the partition is reassigned to another consumer. Now you potentially have two different consumers processing the same message at the same time. If the first consumer hasn't finished processing yet, `isProcessed` will be false for both and they will both continue to process the message. Using something like advisory locks could prevent this case by having whoever comes first take the lock, then check if the message has been processed, process the message and finally write to the DB before releasing the lock. However, what if writing to the DB fails? You can't "unprocess" the message, so it's gonna be at-least once regardless.

The docs should be amended, because while it's true that you can reduce the chance that a message is processed several times, which is good enough for most use-cases, it's not true that it can give you exactly-once semantics.

**Comment by OneCricketeer** [comment]
> First, what if the group is rebalancing and the partition is reassigned to another consumer. Now you potentially have two different consumers processing the same message at the same time


I'm more familiar with the Java API, so apologies, but is there no rebalance listener to prevent that, or any way to catch the equivalent of WakeupException?

**Comment by Nevon** [comment]
This is an issue of distributed systems - whether there's an API for being notified or not doesn't matter. The node won't know at that time that there is a rebalance ongoing. It will only discover that the next time it communicates with the cluster, which will likely be when it heartbeats next.
